### ISSUE TITLE
rrule bug when start date is month end and frequency is monthly

### ISSUE THREAD

**Issue by broncos168** [dateutil/dateutil#149]
Please bear me. I am new to Python and this is my first time post in github. Thanks.

Code snippet to demo the bug:

```python
from datetime import date
from dateutil import rrule
from dateutil import relativedelta

start_date=date(2014, 12, 31)  # date(2014, 12, 29/30/31) 
end_date=date(2015, 12, 31)
freq = rrule.MONTHLY
interval = 1
dts = list(rrule.rrule(freq, interval=interval, dtstart=start_date, until=end_date))
for dt in dts:
    print(dt.date().isoformat())
```

In this edge case, outputs are not monthly
```
2014-12-31
2015-01-31
2015-03-31
2015-05-31
2015-07-31
2015-08-31
2015-10-31
2015-12-31
```

```python
d1=date(2015, 2, 28)
d2=date(2015, 5, 31)
rel_delta1 = relativedelta.relativedelta(d2, d1)
rel_delta2 = relativedelta.relativedelta(d1, d2)
print(rel_delta1, rel_delta2)
```

Output:
```python
relativedelta(months=+3, days=+3) relativedelta(months=-3)
rel_delta1 is not the negative of rel_delta2, vice verse
```

**Comment by phep** [comment]
Hi,
While the behaviour is admittedly quite disconcerting, I'm not quite sure this is a valid bug since RFC5545 says:
```
Recurrence rules may generate recurrence instances with an invalid
date (e.g., February 30) or nonexistent local time (e.g., 1:30 AM
on a day where the local time is moved forward by an hour at 1:00
AM).  Such recurrence instances MUST be ignored and MUST NOT be
counted as part of the recurrence set.
```
The current behaviour respects the RFC.

Maybe there might be some parameter added to let the user decides which behaviour s/he's expecting (be strict and exclude, or floor/ceil the occurence).
I'm just discovering dateutil so this might have already been implemented somehow.

**Comment by pganssle** [comment]
@phep Good call. I don't think it's necessary to add any sort of user option for this. If I recall correctly, if you want to get the last day of the month, you can use `bymonth=-1`. I'll double check that, though.

**Comment by phep** [comment]
Yes, it is true that one can use rrules to express offset from start or end of month, but this is not necessarily semantically equivalent to the given example.

Let's say you sell some online service to which customers may subscribe on a monthly basis and the terms are computed from day of month to day  of month, i.e. if a customer subscribes the 15th of October, the next payment will occur on the 15th of November.

How to deal with the cases where customers first subscribe on the 29, 30 or 31? Presently we are forced to test this in the code to adjust the term date to the month duration, some way or another. Wouldn't it be nice if we'd just have to write something similar to `rrule(MONTHLY, dtstart=today, adjust_end_of_month=True)`?

**Comment by pganssle** [comment]
@phep I'll have to think about this. I can't imagine that iCal doesn't have this problem and given that they expressly thought about it, I imagine they have a way to handle it.

I'm hesitant to start moving beyond the RFC. It may cause problems with the conversion between rrule and rrulestr.

In the relatively simple situation you describe, I'd probably just not use an RRULE at all:

```python
from dateutil.relativedelta import relativedelta
from datetime import datetime
from itertools import count

dtstart = datetime(2015, 1, 31)
interval = 1
MONTH = relativedelta(months=1)

rr = (dtstart + x * MONTH for x in count(1, interval))

print([next(rr) for x in range(5)])
```

Output:
```
[datetime.datetime(2015, 2, 28, 0, 0),
 datetime.datetime(2015, 3, 31, 0, 0),
 datetime.datetime(2015, 4, 30, 0, 0),
 datetime.datetime(2015, 5, 31, 0, 0),
 datetime.datetime(2015, 6, 30, 0, 0)]
```

That said, obviously you lose the general functionality of an rrule by doing that, so it's worth thinking about expressing the same thing as an RRULE.

**Comment by phep** [comment]
@pganssle, going back to the problem, I think you nailed it:

> I'm hesitant to start moving beyond the RFC

"_Special cases aren't special enough to break the rules._" the zen of Python says.

The RFC is not perfect but diverging from it would just bring less perfection, or, in this case, more uncertainty.

Yet the strangeness reported by @broncos168 might deserve some _nota bene_ in the documentation.

**Comment by pganssle** [comment]
Resolved by #213

**Comment by NcsMA** [comment]
Hi,

I've been trying to execute what @pganssle said, it's works when we put as entry january 31, but if I change it to abril it's bring back the errors on the last day of the months:

![image](https://user-images.githubusercontent.com/89947252/211017728-700397ac-c637-40f6-ac0b-0449bf2cc364.png)

the return is:

![image](https://user-images.githubusercontent.com/89947252/211017820-d87175ff-6a32-466f-a2ec-9e6b198665ec.png)

Can someone help me with that?
